2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,242 @@
+
+### Installing resources
+
+`Install-PSResource` will only work for modules.
+A `-AllowPrerelease` switch allows installing prerelease versions.
+Other types will use `Save-PSResource` (see below).
+A `-Repository` parameter accepts a specific repository name.
+If `-Repository` is not specified and the same resource is found in multiple
+repositories, then the resource is automatically installed quietly from the
+trusted repository with the highest version matching the `-Version` parameter
+(if specified, otherwise newest non-prerelease version unless `-AllowPrerelease`
+is used).
+If there are no trusted repositories matching the query, then the newest version
+fulfilling the query will be prompted to be installed.
+If there are multiple repositories with the same trust level containing the same
+version, the first one is used.
+
+`-AllowUntrusted` can be used to suppress being prompted for untrusted sources.
+`-AllowDifferentPublisher` can be used to suppress being prompted if the publisher
+of the module is different from the currently installed version.
+`-AllowReinstall` can be used to re-install a module.
+
+A `-Quiet` switch will suppress progress information.
+
+### Dependencies and version management
+
+`Install-PSResource` can accept a path to a psd1 file (using `-RequiredModulesFile`)
+or a hashtable (using `-RequiredModules`) where the key is the module name and the
+value is either the required version specified using Nuget version range syntax or
+a hash table where `repository` is set to the URL of the repository and
+`version` contains the [Nuget version range syntax](https://docs.microsoft.com/en-us/nuget/reference/package-versioning#version-ranges-and-wildcards).
+
+```powershell
+Install-PSResource -RequiredModules @{
+  "Configuration" = "[1.3.1,2.0)"
+  "Pester"        = @{
+    version = "[4.4.2,4.7.0]"
+    repository = "https://www.powershellgallery.com"
+  }
+}
+```
+
+### Saving resources
+
+With the removal of PackageManagement, there is still a need to support saving
+arbitrary nupkgs (assemblies) used for scripts.
+
+`Save-PSResource -Type Library` will download nupkgs that have a `lib` folder.
+A `-AllowPrerelease` switch allows saving prerelease versions.
+The dependent native library in `runtimes` matching the current system runtime
+will be copied to the same destination.
+
+This cmdlet has the same parameters as `Install-PSResource` with the addition
+of mandatory `-Path` or `-LiteralPath` parameters.
+
+### Updating resources
+
+`Update-PSResource` will update all resources to most recent minor version by
+default.
+A `-OnlyMinorUpdates` switch will allow only updating to newer minor version.
+If the installed resource is a pre-release, it will automatically update to
+latest prerelease or stable version (if available).
+
+### Publishing resources
+
+`Publish-PSResource` will supporting publishing modules and scripts.
+Modules will be given as a path to the module folder.
```

Comment:
The wording `module folder` here is interesting. Is there intent to support publication of:

- Modules in folders without manifests?
- Raw psm1s?
- Raw DLLs?

This probably goes to @heaths's question, for which there are two possibilities I see:
1. If `Import-Module <thing>` works, `Publish-Module <thing>` works
2. We formalise some concept of a module dev structure, whereby the shape of the development source differs from the shape of the published thing

I think (2) is worth keeping in mind (not so much for automated building, but for things like platform-specific dependencies), but is a huge work item to do properly and is out of scope for this RFC. For now I think there's a simple elegance to being able to rely on (1) and letting module developers write their own build infrastructure.

Reply:
Ah it turns out this is referred to in Alternate Proposals and Considerations

Reply:
It's worth **considering** _ignoring_ the folder name if we know what the module's name is some other way -- but it's really important to remember that modules can have lots of extra manifest files in there (e.g. a default `Configuration.psd1` file, or even the manifests from nested modules?), so you'd have to be **really sure** you were publishing with the right name.

Reply:
I'll clarify in the RFC that we can support both.  I recall that PSGet today requires the module to be loaded which is annoying.